### PR TITLE
feat: Allow deserialization of candid values with unknown types

### PR DESCRIPTION
**Pull request by frederikrothenberger**
# Description

This change introduces `IDL.Unknown` to the candid library to specify that the type of the value is not known. `IDL.Unknown` can be use for deserialisation only.

The deserialisation of `IDL.Unknown` will be based on the type table included within the candid encoded value. This means that all field names will only be available in hashed form. The decoded value has a function `type()` that returns it's actual candid type.

# How Has This Been Tested?

Unit tests have been added.

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by frederikrothenberger** [comment]
Open questions / things I'm unsure about:
* Primitives are boxed, because otherwise the `type()` function cannot be added.
* IDL.Null does not deserialize to `null` but to an empty object with just the `type` function.
* The types do not always exactly correspond with the original definitions due to the structure of reference types. The resulting types are however compatible with the original types.
* I have now not added a disclaimer to discourage usage by other developers. Is this necessary? Imho, `IDL.Unknown` should never break, as long as Candid remains self-describing (which it hopefully will stay).

@nomeata : For some reason, I cannot add you as a reviewer. That's why I'm pinging you this way.
I implemented this in order to be able to implement https://github.com/dfinity/interface-spec/pull/20 in the service worker (specifically parsing `/* application-specific type */`). What's your opinion on this?

**Comment by frederikrothenberger** [comment]
> Doesn’t that mean it can’t be used for the token, because you need to serialize that again?

The idea is to use `IDL.Unknown` for deserialization. This will yield some value for token. This value will have a function `type()` which returns the actual type that `IDL.Unknown` turned out to be. This type should be used to serialize the value again.

Take a look at the unit test, they should illustrate how to do this.

**Comment by nomeata** [comment]
Ah, I see. Values of this type _can_ be used for serialization (`IDL.encode([value.type()], [value])`) but they can only be _created_ by deserialization. So you can’t (or shouldn’t) convert other types and values to `IDL.Unknown`. Is that correct? (I don’t really know the JS library, so I may be asking stupid questions).

**Comment by nomeata** [comment]
Yeah, then it indeed looks like a reasonable best effort approach.

**Comment by chenyan-dfinity** [comment]
Generally LGTM. My sentiment is the same as @nomeata that this can only be a best effort support from Candid side, as this is not properly spec'ed.

> * Primitives are boxed, because otherwise the `type()` function cannot be added.
> * `IDL.Null` does not deserialize to `null` but to an empty object with just the `type` function.
> * The types do not always exactly correspond with the original definitions due to the structure of reference types. The resulting types are however compatible with the original types.

As it's not in the spec, you are free to choose the representation.

> * I did not add a disclaimer to discourage usage by other developers. Is this ok? Imho, `IDL.Unknown` should never break, as long as Candid remains self-describing (which it hopefully will stay).

Please do add a comment that this is for internal use only. Once we have proper support for generics, we want to migrate away from `Unknown`. Even before that, we don't want to encourage the use of `Unknown` unless you have no choice.

**Review by chenyan-dfinity — APPROVED** [review]

File: docs/generated/changelog.html [hunk 1]
```diff
@@ -28,6 +28,10 @@ <h2>Version 0.10.5</h2>
           versions to 0 for major version updates
         </li>
         <li>Removes jest-expect-message, which was making test error messages less useful</li>
+        <li>
+          Candid now allows decoding values with unknown types using 'IDL.Unknown' as a type
```
[the remark this answers is not in the document]
Comment (krpeacock) [reply docs/generated/changelog.html:32]: I'm okay with leaving it in. If someone is capable of diving into using this feature with zero documentation, I don't want to get in their way of building whatever they want on the IC

File: packages/candid/src/idl.ts [hunk 2]
```diff
@@ -1598,6 +1669,7 @@ export type InterfaceFactory = (idl: {
 // Export Types instances.
 export const Empty = new EmptyClass();
 export const Reserved = new ReservedClass();
+export const Unknown = new UnknownClass();
```
Comment (krpeacock) [inline packages/candid/src/idl.ts:1672]:
```suggestion
/**
 * Client-only type for deserializing unknown data. Not supported by Candid, and its use is discouraged.
 */
export const Unknown = new UnknownClass();
```

**Comment by nomeata** [comment]
Technically it should be possible to represent a value of `Unknown` by the concrete byte sequence that you read from the incoming message, and then just put that into the output stream as is. Not sure if that'a easier or harder to implement.

And you still have to decose the type table, of course, that'd be the same

File: docs/generated/changelog.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenyan-dfinity) [reply docs/generated/changelog.html:32]: My concern is that it's too easy to use. People can use `Unknown` to decode any message and thus sidestep the backward compatibility check.

**Pull request merged**